R/superPartition.R

```diff
@@ -0,0 +1,268 @@
+#'
+#' It may be the case that large super-partitions cannot be easily broken with Genie due to high similarity between features. In this case, we use k-means to break the cluster.
+#'
+#' @examples
+#'
+#' set.seed(123)
+#' df <- simulate_block_data(c(15, 20, 10), lower_corr = .4, upper_corr = .6, n = 100)
+#'
+#' #  don't accept reductions where information < .6
+#' prt <- super_partition(df, threshold = .6, cluster_size = 30)
+#' prt
+#'
+#' @references
+#' Barrett, Malcolm and Joshua Millstein (2020). partition: A fast and flexible framework for data reduction in R. Journal of Open Source Software, 5(47), 1991, https://doi.org/10.21105/joss.01991Millstein J, Battaglin F, Barrett M, Cao S, Zhang W, Stintzing S, et al. Partition: a surjective mapping approach for dimensionality reduction. *Bioinformatics* **36** (2019) 676–681. doi:10.1093/bioinformatics/btz661.
+#'
+#' Gagolewski, Marek, Maciej Bartoszuk, and Anna Cena. "Genie: A new, fast, and outlier-resistant hierarchical clustering algorithm." Information Sciences 363 (2016): 8-23.
+#'
+#' Millstein, Joshua, Francesca Battaglin, Malcolm Barrett, Shu Cao, Wu Zhang, Sebastian Stintzing, Volker Heinemann, and Heinz-Josef Lenz. 2020. “Partition: A Surjective Mapping Approach for Dimensionality Reduction.” *Bioinformatics* 36 (3): https://doi.org/676–81.10.1093/bioinformatics/btz661.
+#'
+#' @seealso [partition()]
+#'
+#' @author Katelyn Queen, \email{kjqueen@@usc.edu}
+#'
+#' @export
+#' @import progress
+#' @import genieclust
+#' @import gtools
+#' @import tibble
+super_partition <- function(full_data,
+                            threshold = 0.5,
+                            cluster_size = 4000,
+                            partitioner = part_icc(),
+                            tolerance = .0001,
+                            niter = NULL,
+                            x = "reduced_var",
+                            .sep = "_") {
+
+  # ensure 0 < threshold < 1
+  if(0 > threshold | 1 < threshold) stop("Threshold must be between 0 and 1.")
+
+  # ensure data frame structure
+  full_data <- as.data.frame(full_data)
+
+  # if < 4,000 features, call regular partition
+  if(ncol(full_data) < cluster_size) {
+    message(paste0("Using regular Partition function since there are < ", cluster_size, "features."))
+    return(partition(full_data, threshold = threshold))
+  }
+
+  # iteration counters
+  i = j = k = l = 0
+
+  # Function to go from cluster-specific column numbers to full data column numbers
+  ## full_data  - sample by probe expression data for all probes
+  ## small_data - sample by probe expression data for probes in specific cluster
+  ## modules   - indices from partition mapping key
+  ## return - indices from partition mapping key for full data
+  full_data_col_numbers <- function(full_data, small_data, modules) {
+
+    # vector of lists to return
+    return_mods <- numeric(length(modules))
+
+    for (i in 1:length(modules)) {
```

Comment:
```suggestion
    for (i in seq_along(modules)) {
```

Reply:
`seq_along()` is usually a better choice for creating ranges based on length as it avoids bugs related to empty objects